gooddata-pandas/gooddata_pandas/result_convertor.py

```diff
@@ -0,0 +1,309 @@
+# (C) 2022 GoodData Corporation
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union, cast
+
+import pandas
+from attrs import Factory, define, field, frozen
+
+from gooddata_sdk import BareExecutionResponse, ExecutionResult
+
+_DEFAULT_PAGE_SIZE = 100
+_DataHeaders = List[List[Any]]
+_DataArray = List[Union[int, None]]
+LabelOverrides = Dict[str, Dict[str, Dict[str, str]]]
+
+
+@frozen
+class _DataWithHeaders:
+    data: Union[_DataArray, List[_DataArray]]
+    """extracted data; either array of values for one-dimensional result or array of arrays of values"""
+
+    data_headers: Tuple[_DataHeaders, Optional[_DataHeaders]]
+    """per-dimension headers for the data"""
+
+    grand_totals: Tuple[Optional[List[_DataArray]], Optional[List[_DataArray]]]
+    """per-dimension grand total data"""
+
+    grand_total_headers: Tuple[Optional[_DataHeaders], Optional[_DataHeaders]]
+    """per-dimension grand total headers"""
+
+
+@define
+class _AccumulatedData:
+    """
+    Utility class to offload code from the function that extracts all data and headers for a
+    particular paged result. The method drives the paging and calls out to this class to accumulate
+    the essential data and headers from the page.
+    """
+
+    data: Any = field(init=False, default=Factory(list))
+    data_headers: List[Optional[_DataHeaders]] = field(init=False, default=Factory(lambda: [None, None]))
+    grand_totals: List[Optional[List[_DataArray]]] = field(init=False, default=Factory(lambda: [None, None]))
+    grand_totals_headers: List[Optional[_DataHeaders]] = field(init=False, default=Factory(lambda: [None, None]))
```

Comment:
Factory has to stay otherwise the default becomes shared across other instances.
More here https://stackoverflow.com/questions/44141104/when-and-why-should-i-use-attr-factory or https://web.archive.org/web/20200221224620id_/http://effbot.org/zone/default-values.htm.

Reply:
Sorry I made a mistake. I meant:

data_headers: List[Optional[_DataHeaders]] = field(init=False, factory=lambda: [None, None])

Using this we can get rid of Factory class import.

Reply:
fixed in last commit